### PR TITLE
Refactor Inducing Point Selection to Use Allocator Classes for Enhanced Flexibility (#377)

### PR DESCRIPTION
**Pull request by yalsaffar**
# Description
This PR proposes a solution to issue #378 . Although the implementation is fully functional and tested, it is presented as an overall proposal, open for discussion and potential refinement.

### Key Changes
- **Refactor of `select_inducing_points` Function**:
  - **Previous Implementation**: Accepted `method` as a string (`"pivoted_chol"`, `"kmeans++"`, `"auto"`, `"sobol"`) and used conditional logic to select inducing points.
  - **New Implementation**: Now accepts an `InducingPointAllocator` instance, with `AutoAllocator` as the default if `allocator` is `None`. This approach allows users to directly pass allocator instances, aligning with the issue’s goal to enable flexible use of custom allocators like `GreedyImprovementReduction`.

- **New `inducing_point_allocators.py` File**:
  - Introduces classes `SobolAllocator`, `KMeansAllocator`, and `AutoAllocator`, all implementing the `InducingPointAllocator` interface from Botorch. This modularizes allocator logic, moving it out of `select_inducing_points` while following the established base class structure.

- **Modifications to Models and Example Files**:
  - Updated `gp_classification.py`, `monotonic_projection_gp.py`, `monotonic_rejection_gp.py`, `semi_p.py`, and `example_problems.py` to handle allocator class instances rather than string-based methods, improving overall consistency.
  - Added imports for the new allocator classes in `__init__.py` for cross-codebase accessibility.

- **Updated Tests**:
  - Adjusted tests in `test_semi_p.py`, `test_utils.py`, and `test_config.py` to work with allocator classes instead of the previous string-based structure.

### Additional Notes
This PR preserves most of the existing logic in `select_inducing_points` to keep changes minimal. I know further work is needed to confirm compatibility with additional Botorch allocators and to support advanced configurations using `from_config` for custom allocator setups. I’d love to hear your feedback on the overall approach before moving forward with these additional refinements.

### CODE REVIEW

**Review by crasanders — CHANGES_REQUESTED** [review]
Looks good other than a few nits. Our internal tests indicated merge conflicts, so you probably need to rebase.

File: aepsych/models/utils.py [hunk 1]
```diff
@@ -54,53 +55,55 @@ def compute_p_quantile(
...
+    """
+    # Ensure allocator is provided, otherwise use AutoAllocator
+    if allocator is None:
+        allocator = AutoAllocator()
+    
+    
+    # Ensure required inputs are provided for certain allocators
+    if isinstance(allocator, SobolAllocator):
+        assert bounds is not None, "Bounds must be provided for Sobol allocator!"
+    elif isinstance(allocator, (GreedyVarianceReduction, AutoAllocator)):
```
[the remark this answers is not in the document]
Comment (yalsaffar) [reply aepsych/models/utils.py:86]: True, will fix that.

File: tests/test_config.py [hunk 2]
```diff
@@ -1017,7 +1018,7 @@ def test_semip_config(self):
         self.assertTrue(model.dim == 2)
         self.assertTrue(model.inducing_size == 10)
         self.assertTrue(model.stim_dim == 1)
-        self.assertTrue(model.inducing_point_method == "sobol")
+        self.assertTrue(model.inducing_point_method == SobolAllocator)
```
Comment (crasanders) [inline tests/test_config.py:1021]: Let's make a new file for InducingPointAllocator tests and check that each class can be correctly read from a config with its arguments (if any).

File: aepsych/models/inducing_point_allocators.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        inducing_points = draw_sobol_samples(bounds=bounds, n=num_inducing, q=1).squeeze()
+
+        # Ensure correct shape in case Sobol sampling returns a 1D tensor
+        if inducing_points.ndim == 1:
+            inducing_points = inducing_points.view(-1, 1)
+
+        return inducing_points   
+    
+    @classmethod
+    def from_config(cls, config:Config) -> 'SobolAllocator':
```
Comment (crasanders) [inline aepsych/models/inducing_point_allocators.py:72]: I thought about this more, and I think we should have the InducingPointAllocators read from configs themselves rather than receive inputs from the model. I think practically what this will look like is we need to register all of our InducingPointAllocators to work with configs, as well as the ones from botorch (see how we do this for botorch's acquisition functions). In the model's `from_config`, we'll get the InducingPointAllocator class and check if it has its own `from_config` method. If it does, we call it, if it doesn't, we can still try to instantiate it without any arguments. We could have some logic to try to infer what arguments should go into the constructor like we do for acquisition functions, but that code is pretty messy, so I don't know if that's a good idea. 

To maintain backwards compatibility, we can check if `inducing_point_method` is a string, and if it is, pass that string to `select_inducing_points`, but add a deprecation warning to that function that advises that you use the InducingPointAllocator classes directly.

**Review by crasanders — CHANGES_REQUESTED** [review]

**Comment by yalsaffar** [comment]
### Changes

1. **Config-Driven Initialization for Allocators**:
   - Updated `inducing_point_allocators` to read parameters directly from the configuration file. Custom allocators (`SobolAllocator`, `KMeansAllocator`, `AutoAllocator`) now have a `from_config` method to allow instantiation with config-defined parameters.
   - For BoTorch allocators without a `from_config` method, the system attempts instantiation without additional parameters(similar to acquisitions).

2. **Enhanced Model `from_config` Handling**:
   - Models now check if an allocator class has a `from_config` method and use it if available, streamlining configuration-based initialization. 
   - Maintains backward compatibility: if `inducing_point_method` is provided as a string, it falls back to using the legacy approach in the `select_inducing_points` method.

3. **Updated `select_inducing_points` Function**:
   - `select_inducing_points` now supports both config-based allocator classes and string-based legacy methods (`"sobol"`, `"auto"`, etc.). When a legacy method is used, a `DeprecationWarning` is raised, guiding users to switch to the allocator classes.

4. **Testing Updates**:
   - Updated tests in `test_semi_p_config` within `test_config` to work with the new allocator structure.
   - Added `test_points_allocators.py` to validate:
     - Config-driven instantiation of allocators.
     - Bounds handling.
     - Legacy method support with deprecation warnings.
     - Compatibility with BoTorch allocators (`GreedyVarianceReduction`, `GreedyImprovementReduction`).

**Comment by JasonKChow** [comment]
A new style for defining initializations from config just went in. Moving forward, we want to ensure all objects initialized to use the new ConfigurableMixin. Can you update to match the new spec?

ConfigurableMixin got updated here:
https://github.com/facebookresearch/aepsych/blob/main/aepsych/config.py#L430

You can see how I implemented it in parameters here:
https://github.com/facebookresearch/aepsych/blob/main/aepsych/transforms/parameters.py#L98

File: aepsych/models/inducing_point_allocators.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        # Ensure inputs are unique to avoid duplication issues with k-means++
+        unique_inputs = torch.unique(inputs, dim=0)
+
+        # If unique inputs are less than or equal to the required inducing points, return them directly
+        if unique_inputs.shape[0] <= num_inducing:
+            return unique_inputs
+
+        # Run k-means++ on the unique inputs to select inducing points
+        inducing_points = torch.tensor(
+            kmeans2(unique_inputs.numpy(), num_inducing, minit="++")[0],
```
Comment (JasonKChow) [inline aepsych/models/inducing_point_allocators.py:152]: `unique_inputs.numpy()` fails on GPU. Need to force it to cpu first.
```suggestion
            kmeans2(unique_inputs.cpu().numpy(), num_inducing, minit="++")[0],
```

**Review by JasonKChow — CHANGES_REQUESTED** [review]

File: aepsych/models/inducing_point_allocators.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+            config (Config): Configuration object.
+            name (str, optional): Name of the allocator, defaults to None.
+            options (Dict[str, Any], optional): Additional options, defaults to None.
+
+        Returns:
+            Dict[str, Any]: Configuration options for the SobolAllocator.
+        """
+        if name is None:
+            name = cls.__name__
+        bounds = config.gettensor(name, "bounds")
```
Comment (crasanders) [inline aepsych/models/inducing_point_allocators.py:109]: I think we usually don't have bounds explicitly in the config, so we should construct them from lb/ub.

**Review by crasanders — CHANGES_REQUESTED** [review]

File: aepsych/models/inducing_point_allocators.py (hunk 5 above)
Comment (JasonKChow) [reply aepsych/models/inducing_point_allocators.py:109]: I think a utility function went in recently that'll produce 2D bounds. It should be in the root utils.py

Comment (JasonKChow) [reply aepsych/models/inducing_point_allocators.py:109]: Still a problem. Find the get_bounds function in transform and use it here. Ensure a test covers any case where this would break if get_bounds moved.

**Review by JasonKChow — CHANGES_REQUESTED** [review]

**Review by JasonKChow — APPROVED** [review]

File: aepsych/models/inducing_point_allocators.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        ).squeeze()
+
+        # Ensure correct shape in case Sobol sampling returns a 1D tensor
+        if inducing_points.ndim == 1:
+            inducing_points = inducing_points.view(-1, 1)
+
+        return inducing_points
+
+    @classmethod
+    def from_config(
```
Comment (crasanders) [inline aepsych/models/inducing_point_allocators.py:73]: I think we can remove this, and all the other `from_config` methods, because these classes can just inherit the base implementation from the mixin, no?

**Review by crasanders — CHANGES_REQUESTED** [review]
See my comment about the `from_config` implementations. I think once that is resolved, we can land this.

**Comment by JasonKChow** [comment]
@yalsaffar Need rebase.

File: aepsych/models/monotonic_rejection_gp.py
Comment (JasonKChow) [inline aepsych/models/monotonic_rejection_gp.py:1]: Weirdly, I think monotonic rejection GP cannot have their inducing point method set via config. Add it to support your new allocators.

File: aepsych/models/semi_p.py [hunk 7]
```diff
@@ -259,7 +261,7 @@ def __init__(
         slope_mean: float = 2,
         inducing_size: Optional[int] = None,
         max_fit_time: Optional[float] = None,
-        inducing_point_method: str = "auto",
+        inducing_point_method: Optional[InducingPointAllocator] = None,
```
Comment (JasonKChow) [inline aepsych/models/semi_p.py:264]: Default change.

File: aepsych/models/semi_p.py [hunk 8]
```diff
@@ -279,11 +281,7 @@ def __init__(
             inducing_size (int, optional): Number of inducing points. Defaults to 99.
             max_fit_time (float, optional): The maximum amount of time, in seconds, to spend fitting the model. If None,
                 there is no limit to the fitting time.
-            inducing_point_method (string): The method to use to select the inducing points. Defaults to "auto".
-                If "sobol", a number of Sobol points equal to inducing_size will be selected.
-                If "pivoted_chol", selects points based on the pivoted Cholesky heuristic.
-                If "kmeans++", selects points by performing kmeans++ clustering on the training data.
-                If "auto", tries to determine the best method automatically.
+            inducing_point_method (InducingPointAllocator, optional): The method to use to select the inducing points. Defaults to AutoAllocator.
```
Comment (JasonKChow) [inline aepsych/models/semi_p.py:284]: Default change.

File: aepsych/models/semi_p.py [hunk 9]
```diff
@@ -315,6 +313,8 @@ def __init__(
         assert isinstance(
             likelihood, LinearBernoulliLikelihood
         ), "SemiP model only supports linear Bernoulli likelihoods!"
+        if inducing_point_method is None:
```
Comment (JasonKChow) [inline aepsych/models/semi_p.py:316]: Default change.

File: aepsych/models/semi_p.py [hunk 10]
```diff
@@ -528,7 +532,7 @@ def __init__(
         slope_mean: float = 2,
         inducing_size: Optional[int] = None,
         max_fit_time: Optional[float] = None,
-        inducing_point_method: str = "auto",
+        inducing_point_method: Optional[InducingPointAllocator] = None,
```
Comment (JasonKChow) [inline aepsych/models/semi_p.py:535]: Default change.

File: aepsych/models/semi_p.py [hunk 11]
```diff
@@ -556,6 +560,8 @@ def __init__(
             optimizer_options (Dict[str, Any], optional): Optimizer options to pass to the SciPy optimizer during
                 fitting. Assumes we are using L-BFGS-B.
         """
+        if inducing_point_method is None:
```
Comment (JasonKChow) [inline aepsych/models/semi_p.py:563]: Default change.

File: aepsych/models/utils.py [hunk 12]
```diff
@@ -61,72 +65,85 @@ def compute_p_quantile(
 
 def select_inducing_points(
```
Comment (JasonKChow) [inline aepsych/models/utils.py:66]: Consider carefully if this whole utility function is needed.  A lot of the logic here should probably be placed in AutoAllocator instead which can create new allocators within itself to manage auto selection.

File: tests/models/test_gp_classification.py [hunk 13]
```diff
@@ -798,10 +798,21 @@ def test_extra_ask_warns(self):
     def test_hyperparam_consistency(self):
         # verify that creating the model `from_config` or with `__init__` has the same hyperparams
 
-        m1 = GPClassificationModel(lb=[1, 2], ub=[3, 4])
+        m1 = GPClassificationModel(lb=[1, 2], ub=[3, 4], inducing_size=2)
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:801]: Why was this needed, does default the default not work? Defaults should be able to take over if the argument is not set.

File: tests/models/test_utils.py [hunk 14]
```diff
@@ -6,11 +6,17 @@
 # LICENSE file in the root directory of this source tree.
 
 import unittest
+from io import IOBase
```
Comment (JasonKChow) [inline tests/models/test_utils.py:9]: Why is this needed?

File: tests/models/test_utils.py [hunk 15]
```diff
@@ -79,24 +85,28 @@ def test_select_inducing_points(self):
...
                     covar_module=model.covar_module,
                     X=model.train_inputs[0],
                     bounds=model.bounds,
-                    method="kmeans++",
                 )
             ),
             20,
         )
 
-        with self.assertRaises(AssertionError):
```
Comment (JasonKChow) [inline tests/models/test_utils.py:92]: Assess whether the select_inducing_points function should e used at all anymore and consider replacing this check with a commensurate check that looks for invalid inputs (possibly for AutoAllocators).

File: tests/test_points_allocators.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+class TestInducingPointAllocators(unittest.TestCase):
+    def test_sobol_allocator_from_config(self):
+        config_str = """
+            [common]
+            parnames = [par1]
+
+            [par1]
+            par_type = continuous
+            lower_bound = 0.0
+            upper_bound = 1.0
```
Comment (JasonKChow) [inline tests/test_points_allocators.py:29]: Check if bounds are correct if a transformation is applied (easiest to check is `log_scale = True`).

File: tests/test_points_allocators.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+from botorch.models import SingleTaskGP
+from botorch.models.utils.inducing_point_allocators import (
+    GreedyImprovementReduction,
+    GreedyVarianceReduction,
+    InducingPointAllocator,
+)
+from botorch.utils.sampling import draw_sobol_samples
+
+
+class TestInducingPointAllocators(unittest.TestCase):
```
Comment (JasonKChow) [inline tests/test_points_allocators.py:20]: Need to test instantiating a full model and checking if the inducing points are correct: initially dummy but with the correct inducing_point_method set. Then after fitting once, inducing points are replaced with real inducing points.

File: tests/test_points_allocators.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+                bounds=bounds,
+            )
+            self.assertEqual(points.shape, (5, 1))
+
+
+class TestGreedyAllocators(unittest.TestCase):
+    def test_greedy_variance_reduction_allocate_inducing_points(self):
+        # Mock data for testing
+        inputs = torch.rand(100, 2)
+        model = SingleTaskGP(inputs, torch.sin(inputs.sum(dim=-1)).unsqueeze(-1))
```
Comment (JasonKChow) [inline tests/test_points_allocators.py:126]: Should test with our models for completeness. See above for testing inducing points in a model. This might be a good place to do that.

**Review by JasonKChow — CHANGES_REQUESTED** [review]

**Review by JasonKChow — COMMENTED** [review]
Unresolved comments.

File: aepsych/models/gp_classification.py [hunk 19]
```diff
@@ -122,7 +125,7 @@ def __init__(
             learn_inducing_locations=False,
         )
         super().__init__(variational_strategy)
-
+        self.variational_strategy = variational_strategy
```
[the remark this answers is not in the document]
Comment (yalsaffar) [reply aepsych/models/gp_classification.py:128]: The main reason for saving this separately was that it is being called in `test_points_allocators.py`, where `model.variational_strategy.inducing_points` is retrieved and used to check if they match.

[the remark this answers is not in the document]
Comment (JasonKChow) [reply aepsych/models/gp_classification.py:128]: The test shouldn't require changing the spec of the model. Change the test and not the model. You can probably also access the objects by looking closely at where the generated inducing points actually go in the modules.

File: tests/test_points_allocators.py (hunk 16 above)
Comment (JasonKChow) [reply tests/test_points_allocators.py:28]: It must be tested here, transforms are almost always applied (the Normalize transform), the log_scale is simply the most obvious when something goes wrong.

**Review by JasonKChow — APPROVED** [review]
